File: app/src/main/java/com/odnovolov/randomuser/navigation/NavigatorImpl.java
package com.odnovolov.randomuser.navigation;

import android.app.Activity;
import android.app.Application.ActivityLifecycleCallbacks;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentManager;

import com.odnovolov.randomuser.MainActivity;
import com.odnovolov.randomuser.R;
import com.odnovolov.randomuser.model.User;
import com.odnovolov.randomuser.model.UserImpl;
import com.odnovolov.randomuser.screen.home.HomeFragment;
import com.odnovolov.randomuser.screen.userprofile.UserProfileFragment;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class NavigatorImpl implements Navigator, ActivityLifecycleCallbacks {
    @Inject
    public NavigatorImpl() {
    }

    private FragmentManager fragmentManager;

    @Override
    public void onActivityCreated(
            @NonNull Activity activity,
            @Nullable Bundle savedInstanceState
    ) {
        if (activity instanceof MainActivity) {
            if (savedInstanceState == null) {
                ((MainActivity) activity).getSupportFragmentManager().beginTransaction()
                        .replace(R.id.fragmentContainer, new HomeFragment())
                        .commit();
            }
        }
    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {
        if (activity instanceof MainActivity) {
            fragmentManager = ((MainActivity) activity).getSupportFragmentManager();
        }
    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {
        if (activity instanceof MainActivity) {
            fragmentManager = null;
        }
    }

    @Override
    public void navigateToUserProfile(User user) {
        if (fragmentManager == null) return;
        UserProfileFragment fragment = UserProfileFragment.create((UserImpl) user);
        fragmentManager.beginTransaction()
                .replace(R.id.fragmentContainer, fragment)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {}

    @Override
    public void onActivityStopped(@NonNull Activity activity) {}

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle outState) {}

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {}
}

File: app/src/main/java/com/odnovolov/randomuser/navigation/NavigatorImpl.kt
package com.odnovolov.randomuser.navigation

import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import androidx.fragment.app.FragmentManager
import com.odnovolov.randomuser.MainActivity
import com.odnovolov.randomuser.R
import com.odnovolov.randomuser.model.User
import com.odnovolov.randomuser.screen.home.HomeFragment
import com.odnovolov.randomuser.screen.userprofile.UserProfileFragment
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NavigatorImpl @Inject constructor() : Navigator, ActivityLifecycleCallbacks {
    private var fragmentManager: FragmentManager? = null

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        if (activity is MainActivity) {
            if (savedInstanceState == null) {
                activity.supportFragmentManager.beginTransaction()
                    .replace(R.id.fragmentContainer, HomeFragment())
                    .commit()
            }
        }
    }

    override fun onActivityResumed(activity: Activity) {
        if (activity is MainActivity) {
            fragmentManager = activity.supportFragmentManager
        }
    }

    override fun onActivityPaused(activity: Activity) {
        if (activity is MainActivity) {
            fragmentManager = null
        }
    }

    override fun navigateToUserProfile(user: User) {
        val fragmentManager: FragmentManager = this.fragmentManager ?: return
        val fragment: UserProfileFragment = UserProfileFragment.create(user)
        fragmentManager.beginTransaction()
            .replace(R.id.fragmentContainer, fragment)
            .addToBackStack(null)
            .commit()
    }

    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
}